File: uPython/a_pc_scripts/a_bencher.py

Lang_name = "uPython"
Freq = 0

class BenchTimer:
    def __init__(self):
        self.tStart = 0
        self.started = False
        self.tDuration = 0
        self.cDuration = 0
        
    def start(self):
        self.tStart = 0
        self.started = True
        
    def stop(self):
        self.tDuration = 0
        self.cDuration = 0
        if self.started != True :
            print("Timer was not started duration is invalid")
        self.started = False
        
##-------------------------------------------------------------------------
class PrintType:
    Readable = 0
    Matlab = 1
    CSV = 2
    
def Print_header(ptype):
    if ptype == PrintType.Readable:
        print("|%10s|%10s|%10s|%10s|%10s|%10s|%10s|%10s" % ("Lang", "Test", "Freq(Mhz)", "Iter", "Length", "CPU cyc", "Time us", "Rez"))
    elif ptype == PrintType.CSV:
        print("%10s;%10s;%10s;%10s;%10s;%10s;%10s;%10s" % ("Lang", "Test", "Freq(Mhz)", "Iter", "Length", "CPU cyc", "Time us", "Rez"))
    elif ptype == PrintType.Matlab:
        print("PrintType MATLAB not implemented\r\n")
    else:
        print("No such printtype\r\n")

##-------------------------------------------------------------------------
class RezVerifcation:
    Match    = "OK"
    NoMatch  = "Err"
    NoStd    = "NoStd"
    
##-------------------------------------------------------------------------
    
class Tester:
    TestName    = ""
    TestLengths = []
    Niterations = 0
    PrintType   = PrintType.Readable
    RunFn       = []
    
    def Run(self):
        timer = BenchTimer()
        for dataLen in self.TestLengths:
            rez = self.RunFn(dataLen, timer)
            for iters in range(self.Niterations):
                rez = self.RunFn(dataLen, timer)
                if self.PrintType == PrintType.Readable:
                    print("|%10s|%10s|%10s|%10s|%10s|%10s|%10s|%10s" % (Lang_name, self.TestName, Freq, iters+1, dataLen, timer.cDuration, timer.tDuration, rez))
                elif self.PrintType == PrintType.CSV:
                    print("%10s;%10s;%10s;%10s;%10s;%10s;%10s;%10s" % (Lang_name, self.TestName, Freq, iters+1, dataLen, timer.cDuration, timer.tDuration, rez))
                elif self.PrintType == PrintType.Matlab:
                    print("PrintType MATLAB not implemented\r\n")
                else:
                    print("No such printtype\r\n")
            
        

        
    

File: rust-std/src/a_bencher.rs
pub mod my_bench{
    use crate::a_data::data;
    use std::{fmt, ffi::c_uint};
    use esp_idf_sys;

    pub struct timer{
        tStart          : i64,
        cStart          : c_uint,
        started         : bool,
        pub tDuration   : i64,
        pub cDuration   : u32,
    }

    impl timer{
        pub fn init() -> timer{
            timer {
                tStart      : 0,
                cStart      : 0,
                started     : false,
                tDuration   : 0,
                cDuration   : 0,
            }
        }
        
        pub fn start(&mut self){
            self.started = true;
            unsafe{
                self.tStart = esp_idf_sys::esp_timer_get_time();
                self.cStart = esp_idf_sys::xthal_get_ccount();
            }
        }
        
        pub fn stop(&mut self){
            unsafe{
                self.tDuration = esp_idf_sys::esp_timer_get_time() - self.tStart;
                self.cDuration = esp_idf_sys::xthal_get_ccount() - self.cStart;
            }
            if !self.started{
                println!("Timer was not started duration is invalid");
            }
            self.started = false;
        }
    }

    //-------------------------------------------------------------------------
    
    pub enum rezVerifcation {
        Match,
        NoMatch,
        NoStd,
    }

    impl fmt::Display for rezVerifcation {
        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
           match *self {
               rezVerifcation::Match => write!(f, "OK"),
               rezVerifcation::NoMatch => write!(f, "Err"),
               rezVerifcation::NoStd => write!(f, "NoStd"),
           }
        }
    }
    
    //------------------------------------------------------------------------
    
    pub enum printType {
        Readable,
        Matlab,
        CSV,
    }
    
    pub fn print_header(ptype : printType){
        match ptype {
            printType::Readable => {
                println!("|{:<10}|{:<10}|{:<10}|{:<10}|{:<10}|{:<10}|{:<10}|{}", "Lang", "Test", "Freq(Mhz)", "Iter", "Length", "CPU cyc", "Time us", "Rez");
            }
            printType::CSV =>{
                println!("{:<10};{:<10};{:<10};{:<10};{:<10};{:<10};{:<10};{}", "Lang", "Test", "Freq(Mhz)", "Iter", "Length", "CPU cyc", "Time us", "Rez");
            }
            printType::Matlab => todo!(),
        }
    }

    //-------------------------------------------------------------------------
    
    pub struct tester{
        pub testName : &'static str,
        pub testLengths : Vec<usize>,
        pub nIterations : u32,
        pub printType : printType,
        pub runFn : fn(usize, &mut timer) -> rezVerifcation,
    }
    
    impl tester{
        pub fn run(&mut self){
            let CPU_freq:u32;
            unsafe{
                CPU_freq = esp_idf_sys::ets_get_cpu_frequency();
            }
            let mut timer = timer::init();
            for i in 0..self.testLengths.len(){
                let data_len = self.testLengths[i];
                let _rez = (self.runFn)(data_len, &mut timer);
                for iters in 0..self.nIterations {
                    let rez = (self.runFn)(data_len, &mut timer);
                    match self.printType {
                        printType::Readable => {
                            println!("|{:<10}|{:<10}|{:<10}|{:<10}|{:<10}|{:<10}|{:<10}|{}", data::LANG_NAME, self.testName, CPU_freq, iters+1, data_len, timer.cDuration, timer.tDuration, rez);
                        }
                        printType::CSV =>{
                            println!("{:<10};{:<10};{:<10};{:<10};{:<10};{:<10};{:<10};{}", data::LANG_NAME, self.testName, CPU_freq, iters+1, data_len, timer.cDuration, timer.tDuration, rez);
                        }
                        printType::Matlab => todo!(),
                    }
                }
            }
        }
    }
}

